pubnub-core/src/data/request.rs

```diff
@@ -0,0 +1,33 @@
+//! Types used by [`crate::Transport`].
+
+use crate::data::object::Object;
+use crate::data::timetoken::Timetoken;
+
+/// A request to publish a message to a channel.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Publish {
+    /// A channel name to publish the message to.
+    pub channel: String,
+    /// The body of the message.
+    pub payload: Object,
+    /// Additional information associated with the message.
+    pub meta: Option<Object>,
+}
+
+/// Subscribe to messages on channels and/or channel groups.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Subscribe {
+    /// The channel name(s) you are subscribing to.
+    /// Verify that channels are comprised of valid characters.
+    /// You may subscribe to mulitple channels using a comma seperator.
+    /// If subscribing to no channels (only channel groups),
+    /// use a comma char (,) as a placeholder.
+    /// You may subscribe to channels, channels & channel groups,
+    /// or just channel groups.
+    pub channels: Vec<String>,
```

Comment:
Yes, this is exactly right. The comment was just copied from the PubNub docs website, and in this context, it's very misleading. I've already corrected it in the next iteration of my work. I'll submit the next PR as soon as we get this one done.
I've also added the type wrapper for `Channel(String)`, but ended up removing it, cause there're really no issues we can encounter if we just use strings at the core - everything is handled by the transport layer, and the channels themselves don't have any documented restrictions - and all the encoding troubles with HTTP protocol `pubnub_hyper` concern.

Reply:
It's actually not legal to have a comma character in a channel name, since this will just confuse the server into seeing two (or more) channels... Empty string is also an invalid name.

You're right, these aren't documented. 😭

Reply:
Hmm, that's right. I'll add it in the next pass - atop of my other changes.